README
      _                _          __ 
     | | version 0.4.1| |        / _|
   __| |_ __  ___  ___| |__   ___| |_ 
  / _` | '_ \/ __|/ __| '_ \ / _ \  _|
 | (_| | | | \__ \ (__| | | |  __/ |  
  \__,_|_| |_|___/\___|_| |_|\___|_|

       D O C U M E N T A T I O N

DNSChef is a highly configurable DNS proxy for Penetration Testers and Malware Analysts. A DNS proxy (aka "Fake DNS") is a tool used for application network traffic analysis among other uses. For example, a DNS proxy can be used to fake requests for "badguy.com" to point to a local machine for termination or interception instead of a real host somewhere on the Internet. 

There are several DNS Proxies out there. Most will simply point all DNS queries a single IP address or implement only rudimentary filtering. DNSChef was developed as part of a penetration test where there was a need for a more configurable system. As a result, DNSChef is cross-platform application capable of forging responses based on inclusive and exclusive domain lists, supporting multiple DNS record types, matching domains with wildcards, proxying true responses for nonmatching domains, defining external configuration files, IPv6 and many other features. You can find detailed explanation of each of the features and suggested uses below.

The use of DNS Proxy is recommended in situations where it is not possible to force an application to use some other proxy server directly. For example, some mobile applications completely ignore OS HTTP Proxy settings. In these cases, the use of a DNS proxy server such as DNSChef will allow you to trick that application into forwarding connections to the desired destination.

Setting up a DNS Proxy
======================

Before you can start using DNSChef, you must configure your machine to use a DNS nameserver with the tool running on it. You have several options based on the operating system you are going to use:

* **Linux** - Edit */etc/resolv.conf* to include a line on the very top with your traffic analysis host (e.g add "nameserver 127.0.0.1"  if you are running locally). Alternatively, you can add a DNS server address using tools such as Network Manager. Inside the Network Manager open IPv4 Settings, select *Automatic (DHCP) addresses only* or *Manual* from the *Method* drop down box and edit *DNS Servers* text box to include an IP address with DNSChef running.

* **Windows** - Select *Network Connections* from the *Control Panel*. Next select one of the connections (e.g. "Local Area Connection"), right-click on it and select properties. From within a newly appearing dialog box, select *Internet Protocol (TCP/IP)* and click on properties. At last select *Use the following DNS server addresses* radio button and enter the IP address with DNSChef running. For example, if running locally enter 127.0.0.1.

* **OS X** - Open *System Preferences* and click on the *Network* icon. Select the active interface and fill in the *DNS Server* field. If you are using Airport then you will have to click on *Advanced...* button and edit DNS servers from there. Alternatively, you can edit */etc/resolv.conf* and add a fake nameserver to the very top there (e.g "nameserver 127.0.0.1").

* **iOS** - Open *Settings* and select *General*. Next select on *Wi-Fi* and click on a blue arrow to the right of an active Access Point from the list. Edit DNS entry to point to the host with DNSChef running. Make sure you have disabled Cellular interface (if available).

* **Android** - Open *Settings* and select *Wireless and network*.  Click on *Wi-Fi settings* and select *Advanced* after pressing the *Options* button on the phone. Enable *Use static IP* checkbox and configure a custom DNS server.

If you do not have the ability to modify device's DNS settings manually, then you still have several options involving techniques such as [ARP Spoofing](http://en.wikipedia.org/wiki/ARP_spoofing), [Rogue DHCP](http://www.yersinia.net/doc.htm) and other creative methods.

At last you need to configure a fake service where DNSChef will point all of the requests. For example, if you are trying to intercept web traffic, you must bring up either a separate web server running on port 80 or set up a web proxy (e.g. Burp) to intercept traffic. DNSChef will point queries to your proxy/server host with properly configured services.

Running DNSChef
===============

DNSChef is a cross-platform application developed in Python which should run on most platforms which have a Python interpreter. You can use the supplied *dnschef.exe* executable to run it on Windows hosts without installing a Python interpreter. This guide will concentrate on Unix environments; however, all of the examples below were tested to work on Windows as well.

Let's get a taste of DNSChef with its most basic monitoring functionality. Execute the following command as root (required to start a server on port 53):

    # ./dnschef.py
        
              _                _          __  
             | | version 0.2  | |        / _| 
           __| |_ __  ___  ___| |__   ___| |_ 
          / _` | '_ \/ __|/ __| '_ \ / _ \  _|
         | (_| | | | \__ \ (__| | | |  __/ |  
          \__,_|_| |_|___/\___|_| |_|\___|_|  
                       iphelix@thesprawl.org  

    [*] DNSChef started on interface: 127.0.0.1 
    [*] Using the following nameservers: 8.8.8.8
    [*] No parameters were specified. Running in full proxy mode


Without any parameters, DNSChef will run in full proxy mode. This means that all requests will simply be forwarded to an upstream DNS server (8.8.8.8 by default) and returned back to the quering host. For example, let's query an "A" record for a domain and observe results:

    $ host -t A thesprawl.org
    thesprawl.org has address 108.59.3.64

DNSChef will print the following log line showing time, source IP address, type of record requested and most importantly which name was queried:

    [23:54:03] 127.0.0.1: proxying the response of type 'A' for thesprawl.org

This mode is useful for simple application monitoring where you need to figure out which domains it uses for its communications.

DNSChef has full support for IPv6 which can be activated using *-6* or *--ipv6** flags. It works exactly as IPv4 mode with the exception that default listening interface is switched to ::1 and default DNS server is switched to 2001:4860:4860::8888. Here is a sample output:

    # ./dnschef.py -6
              _                _          __
             | | version 0.2  | |        / _|
           __| |_ __  ___  ___| |__   ___| |_
          / _` | '_ \/ __|/ __| '_ \ / _ \  _|
         | (_| | | | \__ \ (__| | | |  __/ |
          \__,_|_| |_|___/\___|_| |_|\___|_|
                       iphelix@thesprawl.org

    [*] Using IPv6 mode.
    [*] DNSChef started on interface: ::1
    [*] Using the following nameservers: 2001:4860:4860::8888
    [*] No parameters were specified. Running in full proxy mode
    [00:35:44] ::1: proxying the response of type 'A' for thesprawl.org
    [00:35:44] ::1: proxying the response of type 'AAAA' for thesprawl.org
    [00:35:44] ::1: proxying the response of type 'MX' for thesprawl.org

NOTE: By default, DNSChef creates a UDP listener. You can use TCP instead with the *--tcp* argument discussed later.

Intercept all responses
-----------------------

Now, that you know how to start DNSChef let's configure it to fake all replies to point to 127.0.0.1 using the *--fakeip* parameter:

    # ./dnschef.py --fakeip 127.0.0.1 -q
    [*] DNSChef started on interface: 127.0.0.1 
    [*] Using the following nameservers: 8.8.8.8
    [*] Cooking all A replies to point to 127.0.0.1
    [23:55:57] 127.0.0.1: cooking the response of type 'A' for google.com to 127.0.0.1
    [23:55:57] 127.0.0.1: proxying the response of type 'AAAA' for google.com
    [23:55:57] 127.0.0.1: proxying the response of type 'MX' for google.com

In the above output you an see that DNSChef was configured to proxy all requests to 127.0.0.1. The first line of log at 08:11:23 shows that we have "cooked" the "A" record response to point to 127.0.0.1. However, further requests for 'AAAA' and 'MX' records are simply proxied from a real DNS server. Let's see the output from requesting program:

    $ host google.com localhost
    google.com has address 127.0.0.1
    google.com has IPv6 address 2001:4860:4001:803::1001
    google.com mail is handled by 10 aspmx.l.google.com.
    google.com mail is handled by 40 alt3.aspmx.l.google.com.
    google.com mail is handled by 30 alt2.aspmx.l.google.com.
    google.com mail is handled by 20 alt1.aspmx.l.google.com.
    google.com mail is handled by 50 alt4.aspmx.l.google.com.

As you can see the program was tricked to use 127.0.0.1 for the IPv4 address. However, the information obtained from IPv6 (AAAA) and mail (MX) records appears completely legitimate. The goal of DNSChef is to have the least impact on the correct operation of the program, so if an application relies on a specific mailserver it will correctly obtain one through this proxied request.

Let's fake one more request to illustrate how to target multiple records at the same time:

    # ./dnschef.py --fakeip 127.0.0.1 --fakeipv6 ::1 -q
    [*] DNSChef started on interface: 127.0.0.1 
    [*] Using the following nameservers: 8.8.8.8
    [*] Cooking all A replies to point to 127.0.0.1
    [*] Cooking all AAAA replies to point to ::1
    [00:02:14] 127.0.0.1: cooking the response of type 'A' for google.com to 127.0.0.1
    [00:02:14] 127.0.0.1: cooking the response of type 'AAAA' for google.com to ::1
    [00:02:14] 127.0.0.1: proxying the response of type 'MX' for google.com

In addition to the --fakeip flag, I have now specified --fakeipv6 designed to fake 'AAAA' record queries. Here is an updated program output:

    $ host google.com localhost
    google.com has address 127.0.0.1
    google.com has IPv6 address ::1
    google.com mail is handled by 10 aspmx.l.google.com.
    google.com mail is handled by 40 alt3.aspmx.l.google.com.
    google.com mail is handled by 30 alt2.aspmx.l.google.com.
    google.com mail is handled by 20 alt1.aspmx.l.google.com.
    google.com mail is handled by 50 alt4.aspmx.l.google.com.

Once more all of the records not explicitly overriden by the application were proxied and returned from the real DNS server. However, IPv4 (A) and IPv6 (AAAA) were both faked to point to a local machine.

DNSChef supports multiple record types:

    +--------+--------------+-----------+--------------------------+
    | Record |  Description |Argument   | Example                  |
    +--------+--------------+-----------+--------------------------+
    |  A     | IPv4 address |--fakeip   | --fakeip 192.0.2.1       |
    |  AAAA  | IPv6 address |--fakeipv6 | --fakeipv6 2001:db8::1   |
    |  MX    | Mail server  |--fakemail | --fakemail mail.fake.com |
    |  CNAME | CNAME record |--fakealias| --fakealias www.fake.com |
    |  NS    | Name server  |--fakens   | --fakens ns.fake.com     |
    +--------+--------------+-----------+--------------------------+

NOTE: For usability not all DNS record types are exposed on the command line. Additional records such as PTR, TXT, SOA, etc. can be specified using the --file flag and an appropriate record header. See the [external definitions file](#external-definitions-file) section below for details.

At last let's observe how the application handles queries of type ANY:

    # ./dnschef.py --fakeip 127.0.0.1 --fakeipv6 ::1 --fakemail mail.fake.com --fakealias www.fake.com --fakens ns.fake.com -q
    [*] DNSChef started on interface: 127.0.0.1 
    [*] Using the following nameservers: 8.8.8.8
    [*] Cooking all A replies to point to 127.0.0.1
    [*] Cooking all AAAA replies to point to ::1
    [*] Cooking all MX replies to point to mail.fake.com
    [*] Cooking all CNAME replies to point to www.fake.com
    [*] Cooking all NS replies to point to ns.fake.com
    [00:17:29] 127.0.0.1: cooking the response of type 'ANY' for google.com with all known fake records.

DNS ANY record queries results in DNSChef returning every faked record that it knows about for an applicable domain. Here is the output that the program will see:

    $ host -t ANY google.com localhost
    google.com has address 127.0.0.1
    google.com has IPv6 address ::1
    google.com mail is handled by 10 mail.fake.com.
    google.com is an alias for www.fake.com.
    google.com name server ns.fake.com.

Filtering domains
-----------------

Using the above example, consider you only want to intercept requests for *thesprawl.org* and leave queries to all other domains such as *webfaction.com* without modification. You can use the *--fakedomains* parameter as illustrated below:

    # ./dnschef.py --fakeip 127.0.0.1 --fakedomains thesprawl.org -q
    [*] DNSChef started on interface: 127.0.0.1
    [*] Using the following nameservers: 8.8.8.8  
    [*] Cooking replies to point to 127.0.0.1 matching: thesprawl.org
    [00:23:37] 127.0.0.1: cooking the response of type 'A' for thesprawl.org to 127.0.0.1
    [00:23:52] 127.0.0.1: proxying the response of type 'A' for mx9.webfaction.com

From the above example the request for *thesprawl.org* was faked; however, the request for *mx9.webfaction.com* was left alone. Filtering domains is very useful when you attempt to isolate a single application without breaking the rest.

NOTE: DNSChef will not verify whether the domain exists or not before faking the response. If you have specified a domain it will always resolve to a fake value whether it really exists or not.

Reverse filtering
-----------------

In another situation you may need to fake responses for all requests except a defined list of domains. You can accomplish this task using the *--truedomains* parameter as follows:

    # ./dnschef.py --fakeip 127.0.0.1 --truedomains thesprawl.org,*.webfaction.com -q
    [*] DNSChef started on interface: 127.0.0.1
    [*] Using the following nameservers: 8.8.8.8  
    [*] Cooking replies to point to 127.0.0.1 not matching: *.webfaction.com, thesprawl.org
    [00:27:57] 127.0.0.1: proxying the response of type 'A' for mx9.webfaction.com
    [00:28:05] 127.0.0.1: cooking the response of type 'A' for google.com to 127.0.0.1

There are several things going on in the above example. First notice the use of a wildcard (*). All domains matching *.webfaction.com will be reverse matched and resolved to their true values. The request for 'google.com' returned 127.0.0.1 because it was not on the list of excluded domains.

NOTE: Wildcards are position specific. A mask of type *.thesprawl.org will match www.thesprawl.org but not www.test.thesprawl.org. However, a mask of type *.*.thesprawl.org will match thesprawl.org, www.thesprawl.org and www.test.thesprawl.org.

External definitions file
-------------------------

There may be situations where defining a single fake DNS record for all matching domains may not be sufficient. You can use an external file with a collection of DOMAIN=RECORD pairs defining exactly where you want the request to go.

For example, let create the following definitions file and call it *dnschef.ini*:

    [A]
    *.google.com=192.0.2.1
    thesprawl.org=192.0.2.2
    *.wordpress.*=192.0.2.3

Notice the section header [A], it defines the record type to DNSChef. Now let's carefully observe the output of multiple queries:

    # ./dnschef.py --file dnschef.ini -q
    [*] DNSChef started on interface: 127.0.0.1 
    [*] Using the following nameservers: 8.8.8.8
    [+] Cooking A replies for domain *.google.com with '192.0.2.1'
    [+] Cooking A replies for domain thesprawl.org with '192.0.2.2'
    [+] Cooking A replies for domain *.wordpress.* with '192.0.2.3'
    [00:43:54] 127.0.0.1: cooking the response of type 'A' for google.com to 192.0.2.1
    [00:44:05] 127.0.0.1: cooking the response of type 'A' for www.google.com to 192.0.2.1
    [00:44:19] 127.0.0.1: cooking the response of type 'A' for thesprawl.org to 192.0.2.2
    [00:44:29] 127.0.0.1: proxying the response of type 'A' for www.thesprawl.org
    [00:44:40] 127.0.0.1: cooking the response of type 'A' for www.wordpress.org to 192.0.2.3
    [00:44:51] 127.0.0.1: cooking the response of type 'A' for wordpress.com to 192.0.2.3
    [00:45:02] 127.0.0.1: proxying the response of type 'A' for slashdot.org

Both *google.com* and *www.google.com* matched the *\*.google.com* entry and correctly resolved to *192.0.2.1*. On the other hand *www.thesprawl.org* request was simply proxied instead of being modified. At last all variations of *wordpress.com*, *www.wordpress.org*, etc. matched the *\*.wordpress.\** mask and correctly resolved to *192.0.2.3*. At last an undefined *slashdot.org* query was simply proxied with a real response.


CHANGELOG
Version 0.4.1

* Add support for HTTPS RR

Version 0.4

* Ported to Python 3.6+
* Made everything a bit more PEP8 compliant
* Improved logging
* Removed IPy library (replaced with built-in ipaddress library)

Version 0.3

* Added support for the latest version of the dnslib library - 0.9.3
* Added support for logging. (idea by kafeine)
* Added support for SRV, DNSKEY, and RRSIG records. (idea by mubix)
* Added support for TCP remote nameserver connections. (idea by mubix)
* DNS name matching is now case insensitive.
* Various small bug fixes and performance tweaks.
* Python libraries are no longer bundled with the distribution, but
  compiled in the Windows binary.

Version 0.2.1

* Fixed a Python 2.6 compatibility issue. (thanks Mehran Goudarzi)

Version 0.2

* Added IPv6 support.
* Added AAAA, MX, CNAME, NS, SOA and NAPTR support.
* Added support for ANY queries (returns all known fake records).
* Changed file format to support more DNS record types.
* Added alternative DNS port support (contributed by fnv).
* Added alternative listening port support for the server (contributed by Mark Straver).
* Updated bundled dnslib library to the latest version - 0.8.2.
* Included IPy library for IPv6 support.

Version 0.1

* First public release


dnsMasterChef.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

from optparse import OptionParser, OptionGroup
from configparser import ConfigParser
from dnslib import *
from IPy import IP
import threading
import random
import operator
import time
import socketserver
import socket
import sys
import os
import binascii
import string
import base64
import time
import dns.resolver
import dns.query
import hashlib
import pyasn
import asyncio
import concurrent.futures
from datetime import datetime

# The database to correlate IP with ASN
ip_to_as = "ipasn_201803.dat"
asndb = ""

if os.path.exists(ip_to_as):
    asndb = pyasn.pyasn(ip_to_as)
else:
    print(ip_to_as + " is not there! I need a ip to AS database...")
    exit(0)

# Providers variable definition
Google = dns.resolver.Resolver()
Google.Name = "Google DNS"
Strongarm = dns.resolver.Resolver()
Strongarm.Name = "Strongarm"
Quad9 = dns.resolver.Resolver()
Quad9.Name = "Quad9"
SafeDNS = dns.resolver.Resolver()
SafeDNS.Name = "SafeDNS"
ComodoSecure = dns.resolver.Resolver()
ComodoSecure.Name = "ComodoSecure"
NortonConnectSafe = dns.resolver.Resolver()
NortonConnectSafe.Name = "NortonConnectSafe"

# Setting IP address of each DNS provider
Google.nameservers = ['8.8.8.8', '8.8.4.4']
Google.Sinkhole = '127.0.0.7'
Quad9.nameservers = ['9.9.9.9', '149.112.112.112']
Quad9.Sinkhole = '127.0.0.2'
Strongarm.nameservers = ['54.174.40.213', '52.3.100.184']
Strongarm.Sinkhole = '127.0.0.3'
SafeDNS.nameservers = ['195.46.39.39', '195.46.39.40']
SafeDNS.Sinkhole = '127.0.0.4'
ComodoSecure.nameservers = ['8.26.56.26', '8.20.247.20']
ComodoSecure.Sinkhole = '127.0.0.5'
NortonConnectSafe.nameservers = ['199.85.126.30', '199.85.127.30']
NortonConnectSafe.Sinkhole = '127.0.0.6'

Providers = [Strongarm, NortonConnectSafe, ComodoSecure, Quad9, SafeDNS]
NumberOfProviders = len(Providers)

# Query a provider and verify the answer
async def Query(domain,DnsResolver,asn_baseline,hash_baseline):
	try:		
		#Get the A record for the specified domain with the specified provider
		Answers = DnsResolver.query(domain, "A")		
		
	#Domain did not resolve
	except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):  
		 return [False, DnsResolver]

	#List of returned IP	
	Arecords = []											
	for rdata in Answers:
		Arecords.append(rdata.address)

	#Compare the answer with the baseline to see if record(s) differ			
	if hashlib.md5(str(sorted(Arecords)).encode('utf-8')).hexdigest() != hash_baseline.hexdigest(): 
	
		#Record(s) differ, checking if the first one is in the same BGP AS								    
		if(asndb.lookup(sorted(Arecords)[0])[0] != asn_baseline):									
			 return [False, DnsResolver]
			 
	#Domain is safe		 
	return [True, DnsResolver]

# Creates the parallels tasks
async def main(domain,asn_baseline,hash_baseline):	
	with concurrent.futures.ThreadPoolExecutor(max_workers=NumberOfProviders) as executor:
		tasks = [
			asyncio.ensure_future(Query(domain, Providers[i],asn_baseline,hash_baseline))
			for i in range(NumberOfProviders)
		]
	   
		for IsSafe,provider in await asyncio.gather(*tasks):
			#One DNS provider in the function 'Query' returned False, so the domain is unsafe
			if IsSafe == False:		
				return [False, provider]
			pass
			
		#Function 'Query' never returned False at this point, the domain is safe
		return [True, provider]	

# Create the loop			
def Createloop(domain,asn_baseline,hash_baseline):
	loop = asyncio.new_event_loop()
	asyncio.set_event_loop(loop)
	result = loop.run_until_complete(main(domain,asn_baseline,hash_baseline))
	
	# return is received, let's close the objects
	loop.run_until_complete(loop.shutdown_asyncgens())			
	return result

#Establish a baseline with Google Public DNS and call function "loop"
def launch(domain):	
	hash_baseline = hashlib.md5()
	try:
			#Lookup the 'A' record(s)
			Answers_Google = Google.query(domain, "A") 		
			
	#Domain did not resolve		
	except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):  
			 return [False, Google]
			 
	# Contain the returned A record(s)		 
	Arecords = []											
	for rdata in Answers_Google: 
		Arecords.append(rdata.address)
	
	#Looking the ASN of the first A record (sorted)
	asn_baseline = asndb.lookup(sorted(Arecords)[0])[0]	
	
	#MD5 Fingerprint of the anwser is the sorted list of A record(s)
	#Because of the round-robin often used in replies.
	#Ex. NS1 returns IP X,Y and NS2 returns IP Y,X
	hash_baseline.update(str(sorted(Arecords)).encode('utf-8')) 
	
	return Createloop(domain,asn_baseline,hash_baseline)


# DNSHandler Mixin. The class contains generic functions to parse DNS requests and
# calculate an appropriate response based on user parameters.
class DNSHandler:

    def parse(self, data):
        response = ''
        try:
            # Parse data as DNS
            d = DNSRecord.parse(data)
        except Exception as e:

            print(('[%s] %s: ERROR: %s' % (time.strftime('%H:%M:%S'),
                                           self.client_address[0], 'invalid DNS request')))
            if self.server.log:
                self.server.log.write(
                    '[%s] %s: ERROR: %s\n' %
                    (time.strftime('%d/%b/%Y:%H:%M:%S %z'),
                     self.client_address[0],
                     'invalid DNS request'))
        else:

            # Only Process DNS Queries
            if QR[d.header.qr] == 'QUERY':
                qname = str(d.q.qname)

                # Chop off the last period
                if qname[-1] == '.': qname = qname[:-1]

                qtype = QTYPE[d.q.qtype]

                # Proxy the request
                if qtype not in ['SOA', 'A']:
                    print("Filtering " + qtype + " requests not supported, Forwarding...")
                    print (
                            "[%s] %s: proxying the response of type '%s' for %s" %
                            (time.strftime("%H:%M:%S"), self.client_address[0], qtype, qname))
                    if self.server.log:
                        self.server.log.write(
                            "[%s] %s: proxying the response of type '%s' for %s\n" %
                            (time.strftime("%d/%b/%Y:%H:%M:%S %z"), self.client_address[0], qtype, qname))

                    nameserver_tuple = random.choice(self.server.nameservers).split('#')
                    response = self.proxyrequest(data, *nameserver_tuple)
                else:
                    IsSafe, ProviderName = launch(qname)
                    if IsSafe:
                        print(qname + " is safe, proxying...")
                        nameserver_tuple = random.choice(self.server.nameservers).split('#')
                        response = self.proxyrequest(data, *nameserver_tuple)
                    else:
                        fake_records = dict()
                        fake_record = ProviderName.Sinkhole
                        fake_records[qtype] = qtype

                        # Create a custom response to the query
                        response = DNSRecord(DNSHeader(id=d.header.id, bitmap=d.header.bitmap, qr=1, aa=1, ra=1), q=d.q)

                        if qtype == "SOA":
                            mname, rname, t1, t2, t3, t4, t5 = fake_record.split(" ")
                            times = tuple([int(t) for t in [t1, t2, t3, t4, t5]])

                            # dnslib doesn't like trailing dots
                            if mname[-1] == ".":
                                mname = mname[:-1]
                            if rname[-1] == ".":
                                rname = rname[:-1]

                            response.add_answer(RR(qname, getattr(QTYPE, qtype),
                                                   rdata=RDMAP[qtype](mname, rname, times)))

                        elif qtype == "A":
                            if fake_record[-1] == ".":
                                fake_record = fake_record[:-1]
                            response.add_answer(RR(qname, getattr(QTYPE, qtype),rdata=RDMAP[qtype](fake_record)))

                        response = response.pack()
                        print(qname + ' Spoofing because it is filtered by ' + ProviderName.Name)
        return response

    # Find appropriate ip address to use for a queried name. The function can

    def findnametodns(self, qname, nametodns):
        # Make qname case insensitive

        qname = qname.lower()

        # Split and reverse qname into components for matching.

        qnamelist = qname.split('.')
        qnamelist.reverse()

        # HACK: It is important to search the nametodns dictionary before iterating it so that
        # global matching ['*.*.*.*.*.*.*.*.*.*'] will match last. Use sorting
        # for that.

        for (domain, host) in sorted(iter(list(nametodns.items())),
                                     key=operator.itemgetter(1)):

            # NOTE: It is assumed that domain name was already lowercased
            #       when it was loaded through --file, --fakedomains or --truedomains
            # don't want to waste time lowercasing domains on every request.

            # Split and reverse domain into components for matching

            domain = domain.split('.')
            domain.reverse()

            # Compare domains in reverse.

            for (a, b) in map(None, qnamelist, domain):
                if a != b and b != '*':
                    break
            else:

                # Could be a real IP or False if we are doing reverse matching
                # with 'truedomains'

                return host
        else:
            return False


# Obtain a response from a real DNS server.

    def proxyrequest(
            self,
            request,
            host,
            port='53',
            protocol='udp',
    ):
        reply = None
        try:
            if self.server.ipv6:

                if protocol == 'udp':
                    sock = socket.socket(socket.AF_INET6,
                                         socket.SOCK_DGRAM)
                elif protocol == 'tcp':
                    sock = socket.socket(socket.AF_INET6,
                                         socket.SOCK_STREAM)
            else:

                if protocol == 'udp':
                    sock = socket.socket(socket.AF_INET,
                                         socket.SOCK_DGRAM)
                elif protocol == 'tcp':
                    sock = socket.socket(socket.AF_INET,
                                         socket.SOCK_STREAM)

            sock.settimeout(3.0)

            # Send the proxy request to a randomly chosen DNS server

            if protocol == 'udp':
                sock.sendto(request, (host, int(port)))
                reply = sock.recv(1024)
                sock.close()
            elif protocol == 'tcp':

                sock.connect((host, int(port)))

                # Add length for the TCP request

                length = binascii.unhexlify('%04x' % len(request))
                sock.sendall(length + request)

                # Strip length from the response

                reply = sock.recv(1024)
                reply = reply[2:]

                sock.close()
        except Exception as e:

            print(('[!] Could not proxy request: %s' % e))
        else:
            return reply


# UDP DNS Handler for incoming requests

class UDPHandler(DNSHandler, socketserver.BaseRequestHandler):

    def handle(self):
        (data, socket) = self.request
        response = self.parse(data)

        if response:
            socket.sendto(response, self.client_address)


# TCP DNS Handler for incoming requests

class TCPHandler(DNSHandler, socketserver.BaseRequestHandler):

    def handle(self):
        data = self.request.recv(1024)

        # Remove the addition "length" parameter used in the
        # TCP DNS protocol

        data = data[2:]
        response = self.parse(data)

        if response:
            # Calculate and add the additional "length" parameter
            # used in TCP DNS protocol

            length = binascii.unhexlify('%04x' % len(response))
            self.request.sendall(length + response)


class ThreadedUDPServer(socketserver.ThreadingMixIn,
                        socketserver.UDPServer):

    # Override SocketServer.UDPServer to add extra parameters

    def __init__(
            self,
            server_address,
            RequestHandlerClass,
            nametodns,
            nameservers,
            ipv6,
            log,
    ):
        self.nametodns = nametodns
        self.nameservers = nameservers
        self.ipv6 = ipv6
        self.address_family = \
            (socket.AF_INET6 if self.ipv6 else socket.AF_INET)
        self.log = log

        socketserver.UDPServer.__init__(self, server_address,
                                        RequestHandlerClass)


class ThreadedTCPServer(socketserver.ThreadingMixIn,
                        socketserver.TCPServer):
    # Override default value

    allow_reuse_address = True

    # Override SocketServer.TCPServer to add extra parameters

    def __init__(
            self,
            server_address,
            RequestHandlerClass,
            nametodns,
            nameservers,
            ipv6,
            log,
    ):
        self.nametodns = nametodns
        self.nameservers = nameservers
        self.ipv6 = ipv6
        self.address_family = \
            (socket.AF_INET6 if self.ipv6 else socket.AF_INET)
        self.log = log

        socketserver.TCPServer.__init__(self, server_address,
                                        RequestHandlerClass)


# Initialize and start the DNS Server

def start_cooking(
        interface,
        nametodns,
        nameservers,
        tcp=False,
        ipv6=False,
        port='55',
        logfile=None,
):
    try:

        if logfile:
            log = open(logfile, 'a', 0)
            log.write('[%s] DNSChef is active.\n'
                      % time.strftime('%d/%b/%Y:%H:%M:%S %z'))
        else:
            log = None

        if tcp:
            print('[*] DNSChef is running in TCP mode')
            server = ThreadedTCPServer(
                (interface, int(port)),
                TCPHandler,
                nametodns,
                nameservers,
                ipv6,
                log,
            )
        else:
            server = ThreadedUDPServer(
                (interface, int(port)),
                UDPHandler,
                nametodns,
                nameservers,
                ipv6,
                log,
            )

        # Start a thread with the server -- that thread will then start
        # more threads for each request

        server_thread = threading.Thread(target=server.serve_forever)

        # Exit the server thread when the main thread terminates

        server_thread.daemon = True
        server_thread.start()

        # Loop in the main thread

        while True:
            time.sleep(100)
    except (KeyboardInterrupt, SystemExit):

        if log:
            log.write('[%s] DNSChef is shutting down.\n'
                      % time.strftime('%d/%b/%Y:%H:%M:%S %z'))
            log.close()

        server.shutdown()
        print('[*] DNSChef is shutting down.')
        sys.exit()
    except IOError:

        print('[!] Failed to open log file for writing.')
    except Exception as e:

        print(('[!] Failed to start the server: %s' % e))


if __name__ == '__main__':
    # Parse command line arguments
    parser = OptionParser(usage="dnschef.py [options]:\n")

    rungroup = OptionGroup(parser, "Optional runtime parameters.")
    rungroup.add_option("--logfile", action="store", help="Specify a log file to record all activity")
    rungroup.add_option("-i", "--interface", metavar="127.0.0.1 or ::1", default="127.0.0.1", action="store",
                        help='Define an interface to use for the DNS listener. By default, the tool uses 127.0.0.1 for IPv4 mode and ::1 for IPv6 mode.')
    rungroup.add_option("--nameservers", metavar="208.67.222.222#53 or 208.67.220.220#53",
                        default='208.67.222.222,208.67.220.220', action="store")
    rungroup.add_option("-t", "--tcp", action="store_true", default=False,
                        help="Use TCP DNS proxy instead of the default UDP.")
    rungroup.add_option("-p", "--port", action="store", metavar="5353", default="5353",
                        help='Port number to listen for DNS requests.')
    parser.add_option_group(rungroup)

    (options, args) = parser.parse_args()
    options.ipv6 = False

    # Main storage of domain filters
    # NOTE: RDMAP is a dictionary map of qtype strings to handling classes
    nametodns = dict()
    for qtype in list(RDMAP.keys()):
        nametodns[qtype] = dict()

    # Notify user about alternative listening port
    if options.port != '53':
        print(('[*] Listening on an alternative port %s' % options.port))


    print(('[*] DNSChef started on interface: %s ' % options.interface))

    # Use alternative DNS servers
    if options.nameservers:
        nameservers = options.nameservers.split(',')
        print(('[*] Using the following nameservers: %s'
               % ', '.join(nameservers)))

    print('[*] No parameters were specified. Running in full proxy mode')

    # Launch DNSChef
    start_cooking(
        interface=options.interface,
        nametodns=nametodns,
        nameservers=nameservers,
        tcp=options.tcp,
        ipv6=options.ipv6,
        port=options.port,
        logfile=options.logfile,
    )
